runtime/compiler/ilgen/Walker.cpp

```diff
@@ -5061,11 +5061,48 @@ TR_J9ByteCodeIlGenerator::loadAuto(TR::DataType type, int32_t slot, bool isAdjun
    push(load);
    }
 
+/**
+ * @brief Returns whether a field ref in the constant pool resolved
+ *
+ * Importantly, when this function returns false, a ResolveCHK is guarenteed to be needed.
```

Comment:
Why do we do runtime resolve when we can safely compile-time resolve it? Other fields only need resolve check when we can't compile-time resolve them. Are ValueType fields different?

Reply:
My understanding was that a resolve check is still needed for compile-time resolved fields. @andrewcraik can you confirm what the correct behaviour is?

Reply:
A compile-time resolve is not a true resolve - it is a promise from the VM that we know what the entry is going to resolve to - if the entry in the constant pool is not resolved when we start compiling and isn't genuinely resolved after the helper call the resolvechk is needed to make the resolution happen at the appropriate point at runtime.

Reply:
@andrewcraik My understanding is that, if VM knows what the entry is going to resolve to, we'll treat it resolved even if the entry is not resolved in cp. That's what we do for other fields (we also do so for method calls except when rtResolve is on).

My impression is that when the JIT ask the VM to compile time resolve something, the VM looks up that thing with restrictions, such as no java code is allowed during the look up. So if a cp entry can be compile-time resolved, it means its resolution has no side effect, and the runtime resolution is not necessary since we know what it is.

While my impression may be inaccurate, my question is with us treating other compile-time resolvable fields resolved, what makes the ValueType fields different? If there is a difference, may be a comment is needed to explain the difference.

Reply:
I don't think there is a difference, but we should ask @tajila or @DanHeidinga to confirm if they want the JIT to call to resolve the CP entry or not or if we can skip that.

Reply:
For fields, if the compile-time resolve is successful then we update the linkage state for the cpEntry. This is equivalent to what would happen in the case of a runtime resolve. If the compile-time resolve fails then the runtime resolution must be run, so that appropriate exceptions can be thrown.

Reply:
Ok so that means for a field resolve that succeeds at compile time we can skip the ResolveCHK, but failure means the ResolveCHK must remain.

Reply:
Changed in force push to return the result of the compile-time resolve.